File: backend/app/api/v1/endpoints/files.py
"""Эндпоинты /files.

Загрузка файлов идёт через presigned POST в S3 — браузер льёт напрямую в
Yandex Object Storage, минуя бэкенд. Бэк только подтверждает наличие и хранит
метаданные.
"""
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, Query, status
from fastapi.responses import Response
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.errors import ApiError
from app.db.session import get_db
from app.integrations.s3 import S3Adapter, get_s3_adapter
from app.modules.auth.dependencies import get_current_user
from app.modules.auth.schemas import OkResponse
from app.modules.files import service
from app.modules.files.pdf_service import render_pdf_from_html
from app.modules.files.models import File, FileEntityType
from app.modules.files.schemas import (
    ConfirmRequest,
    DownloadResponse,
    FileResponse,
    PresignRequest,
    PresignResponse,
    RenderPdfRequest,
)
from app.modules.users.models import User

router = APIRouter(prefix="/files", tags=["files"])


def _s3_dep() -> S3Adapter:
    return get_s3_adapter()


def _to_dto(f: File) -> FileResponse:
    return FileResponse.model_validate(f)


def _pdf_error_details(exc: Exception) -> dict[str, str]:
    raw_message = str(exc).strip()
    first_line = raw_message.splitlines()[0] if raw_message else exc.__class__.__name__
    details: dict[str, str] = {
        "errorType": exc.__class__.__name__,
        "errorMessage": first_line[:300],
    }
    lower = raw_message.lower()
    if "playwright" in lower and ("executable doesn't exist" in lower or "playwright install" in lower):
        details["hint"] = "Run `python -m playwright install chromium` on backend host."
    return details


@router.post("/presign", response_model=PresignResponse, summary="Получить presigned URL для загрузки")
async def presign(
    payload: PresignRequest,
    _: User = Depends(get_current_user),
    s3: S3Adapter = Depends(_s3_dep),
) -> PresignResponse:
    url, fields, file_key, max_bytes = await service.presign(s3, payload)
    return PresignResponse(url=url, fields=fields, file_key=file_key, max_bytes=max_bytes)


@router.post("/confirm", response_model=FileResponse, summary="Подтвердить загрузку файла")
async def confirm(
    payload: ConfirmRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> FileResponse:
    rec = await service.confirm(db, user, payload)
    return _to_dto(rec)


@router.get("", response_model=list[FileResponse], summary="Файлы сущности")
async def list_files(
    entity_type: FileEntityType = Query(alias="entityType"),
    entity_id: uuid.UUID = Query(alias="entityId"),
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[FileResponse]:
    rows = await service.list_for_entity(db, entity_type, entity_id)
    return [_to_dto(r) for r in rows]


@router.get(
    "/{file_id}/download",
    response_model=DownloadResponse,
    summary="Временный URL для скачивания",
)
async def get_download_url(
    file_id: uuid.UUID,
    _: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    s3: S3Adapter = Depends(_s3_dep),
) -> DownloadResponse:
    file = await service.get_file(db, file_id)
    url = await service.download_url(s3, file)
    return DownloadResponse(url=url, expires_in=300)


@router.delete("/{file_id}", response_model=OkResponse, summary="Удалить файл")
async def delete_file(
    file_id: uuid.UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    s3: S3Adapter = Depends(_s3_dep),
) -> OkResponse:
    await service.delete_file(db, s3, user, file_id)
    return OkResponse()


@router.post(
    "/render-pdf",
    summary="Собрать PDF из HTML",
)
async def render_pdf(
    payload: RenderPdfRequest,
    _: User = Depends(get_current_user),
) -> Response:
    try:
        pdf_bytes = await render_pdf_from_html(payload.html)
    except Exception as exc:
        raise ApiError(
            status.HTTP_500_INTERNAL_SERVER_ERROR,
            "pdf_render_failed",
            "Не удалось сформировать PDF на сервере",
            details=_pdf_error_details(exc),
        ) from exc

    safe_name = payload.filename.replace('"', "").replace("\n", "").replace("\r", "")
    return Response(
        content=pdf_bytes,
        media_type="application/pdf",
        headers={"Content-Disposition": f'attachment; filename="{safe_name}"'},
    )

File: frontend/src/api/files.ts
// Тонкий клиент к /files/*.
// MSW для этого домена ничего не делает (см. mocks/handlers.ts) — файлы появились
// сразу с боевым бэком на Этапе 6. Если фронт работает с VITE_USE_MOCKS=true,
// этот API будет уходить «в никуда» и UI должен это учитывать (отрисовка пустого
// списка/ошибки, либо целиком прятать блок «Файлы резюме»).

import { api } from './client';
import type { UUID } from './types';

export type FileEntityType =
  | 'candidate'
  | 'vacancy'
  | 'client'
  | 'contact'
  | 'chat_message'
  | 'document';

export type ScanStatus = 'pending' | 'clean' | 'infected' | 'error';

export interface FileResponse {
  id: UUID;
  ownerUserId: UUID;
  entityType: FileEntityType;
  entityId: UUID;
  fileKey: string;
  originalName: string;
  mime: string;
  size: number;
  scanStatus: ScanStatus;
  scannedAt?: string | null;
  createdAt: string;
}

export interface PresignRequest {
  entityType: FileEntityType;
  entityId: UUID;
  originalName: string;
  mime: string;
  size: number;
}

export interface PresignResponse {
  url: string;
  fields: Record<string, string>;
  fileKey: string;
  maxBytes: number;
}

export interface ConfirmRequest {
  fileKey: string;
  entityType: FileEntityType;
  entityId: UUID;
  originalName: string;
  mime: string;
  size: number;
}

export interface DownloadResponse {
  url: string;
  expiresIn: number;
}

export interface RenderPdfRequest {
  html: string;
  filename: string;
}

export const filesApi = {
  presign: (payload: PresignRequest) =>
    api.post('files/presign', { json: payload }).json<PresignResponse>(),
  confirm: (payload: ConfirmRequest) =>
    api.post('files/confirm', { json: payload }).json<FileResponse>(),
  list: (entityType: FileEntityType, entityId: UUID) =>
    api
      .get('files', { searchParams: { entityType, entityId } })
      .json<FileResponse[]>(),
  download: (id: UUID) => api.get(`files/${id}/download`).json<DownloadResponse>(),
  renderPdf: (payload: RenderPdfRequest) => api.post('files/render-pdf', { json: payload }).blob(),
  remove: (id: UUID) => api.delete(`files/${id}`).json<{ ok: true }>(),
};

/**
 * Хелпер: загружает файл напрямую в S3 через presigned POST.
 * Возвращает созданную запись в БД (через /files/confirm).
 *
 * Использование:
 *   const rec = await uploadFile({ entityType: 'candidate', entityId, file })
 *
 * При желании можно прокинуть `onProgress` через XMLHttpRequest вместо fetch.
 */
export async function uploadFile(args: {
  entityType: FileEntityType;
  entityId: UUID;
  file: File;
}): Promise<FileResponse> {
  const { entityType, entityId, file } = args;
  const pres = await filesApi.presign({
    entityType,
    entityId,
    originalName: file.name,
    mime: file.type || 'application/octet-stream',
    size: file.size,
  });

  // S3 POST: все поля из pres.fields + 'file' в конце.
  const form = new FormData();
  for (const [k, v] of Object.entries(pres.fields)) form.append(k, v);
  form.append('file', file);
  const s3res = await fetch(pres.url, { method: 'POST', body: form });
  if (!s3res.ok) {
    throw new Error(`S3 upload failed: ${s3res.status} ${s3res.statusText}`);
  }

  return filesApi.confirm({
    fileKey: pres.fileKey,
    entityType,
    entityId,
    originalName: file.name,
    mime: file.type || 'application/octet-stream',
    size: file.size,
  });
}
